packages/core/diagnostic/src/diagnostic.js

```diff
@@ -0,0 +1,99 @@
+// @flow
+
+export type DiagnosticHighlightLocation = {|
+  line: number,
+  column: number
+|};
+
+export type DiagnosticSeverity = 'error' | 'warn' | 'info';
+
+export type DiagnosticCodeHighlight = {|
+  start: DiagnosticHighlightLocation,
+  end: DiagnosticHighlightLocation,
+  message?: string
+|};
+
+export type DiagnosticCodeFrame = {|
+  code: string,
+  codeHighlights: DiagnosticCodeHighlight | Array<DiagnosticCodeHighlight>
+|};
+
+// A Diagnostic is a style agnostic way of emitting errors, warnings and info
+// The reporter's are responsible for rendering the message, codeframes, hints, ...
+export type Diagnostic = {|
+  message: string,
+  //? severity: DiagnosticSeverity, // Might add this back later...
+  origin: string, // Name of plugin or file that threw this error
+
+  // Asset metadata
+  filename?: string,
+  language?: string,
+
+  // Codeframe data
+  codeframe?: DiagnosticCodeFrame,
+
+  // Hints to resolve issues faster
+  hints?: string | Array<string>,
+
+  //! Should only be used if there's no way to supply code and codeHighlight...
+  stack?: string
+|};
+
+// This type should represent all error formats Parcel can encounter...
+type PrintableError = Error & {
+  fileName: ?string,
+  codeFrame?: string,
+  highlightedCodeFrame?: string,
+  loc?: {
+    column: number,
+    line: number,
+    ...
+  },
+  source?: string,
+  ...
+};
+
+export function errorToDiagnostic(error: PrintableError): Diagnostic {
+  let codeframe: DiagnosticCodeFrame | void = undefined;
+
+  if (error.loc && error.source) {
+    codeframe = {
+      code: error.source,
+      codeHighlights: {
+        start: {
+          line: error.loc.line,
+          column: error.loc.column
+        },
+        end: {
+          line: error.loc.line,
+          column: error.loc.column
+        }
+      }
+    };
+  }
+
+  return {
+    origin: 'Error',
```

Comment:
Should this go up the tree if filename is defined to find the nearest pkg.json and use the `name` of that package as origin?

Reply:
Is this meant to be the parcel plugin that created the error? If we put a try..catch around all the places we call into plugins in core and wrapped their errors, we could populate this automatically.

Reply:
Yes, this would be the origin of the error so the plugin/parcel pkg that triggers it